### ISSUE TITLE
Rule ST09 not working correctly with Jinja templating

### ISSUE THREAD

**Issue by daczarne** [sqlfluff/sqlfluff#5506]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Rule ST09 is not able to properly asses order of tables when Jinja parameters are present.

### Expected Behaviour

For the order of tables to be assessed correctly.

#### False positive

This example passes when it should fail as the last join sets `ON t3.id = t1.id`.

```sql
SELECT
  t1.a
  , t1.b
  , t2.c
FROM `{{ params.project_id }}.my_dataset.table_1` AS t1
LEFT JOIN `{{ params.project_id }}.my_dataset.table_2` AS t2
  ON t1.id = t2.id
LEFT JOIN `{{ params.project_id }}.my_dataset.table_3` AS t3
  ON t3.id = t1.id
```

![image](https://github.com/sqlfluff/sqlfluff/assets/16988051/66f0790a-f820-45f3-a9e9-e4c4da0bb01a)

#### Expected failure

If we now replace the Jinja params for their values, the check fails as expected.

```sql
SELECT
  t1.a
  , t1.b
  , t2.c
FROM `my_project.my_dataset.table_1` AS t1
LEFT JOIN `my_project.my_dataset.table_2` AS t2
  ON t1.id = t2.id
LEFT JOIN `my_project.my_dataset.table_3` AS t3
  ON t3.id = t1.id
```

![image](https://github.com/sqlfluff/sqlfluff/assets/16988051/2b0d04c8-c7fd-4b37-9c81-324b500ad175)

### How to reproduce

The command I've run is

```shell
sqlfluff lint --config .sqlfluff dags/sql/sqlfluff_test.sql   
```

### Dialect

bigquery

### Version

![image](https://github.com/sqlfluff/sqlfluff/assets/16988051/8b1efc31-b836-4d41-b3a8-b7a6f12edcea)

### Configuration

```cfg
[sqlfluff]
dialect = bigquery
disable_noqa = False
encoding = utf-8
fix_even_unparsable = False
# ignore = templating, parsing
ignore_templated_areas = True
large_file_skip_byte_limit = 0
large_file_skip_char_limit = 0
max_line_length = 120
nocolor = False
output_line_length = 80
processes = 1
runaway_limit = 10
sql_file_exts = .sql
templater = jinja
verbose = 2
warn_unused_ignores = False

rules = ST09

[sqlfluff:templater:jinja]
apply_dbt_builtins = False

[sqlfluff:templater:jinja:context]
params = {"project_id": "my_project"}

[sqlfluff:rules:structure.join_condition_order]
preferred_first_table_in_join_clause = earlier
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
